### ISSUE TITLE
`clippy::not_unsafe_ptr_arg_deref` is right about some functions that should be `unsafe`

### ISSUE THREAD

**Issue by MeGaGiGaGon** [Indra-db/Flecs-Rust#248]
I noticed this while reading though the `World` code, but there are a decent amount of safe functions with `#[allow(clippy::not_unsafe_ptr_arg_deref)]` that probably should be unsafe.

As a side note before I go into all of them, `clippy::not_unsafe_ptr_arg_deref` has a very storied issue [here](https://github.com/rust-lang/rust-clippy/issues/3045) on whether or not these are false positives. I generally agree with Manishearth that they are not, since even though your code might not explicitly dereference the pointer it can easily pass it to another unsafe function/ffi call without doing all the required checks, creating a safety/UB issue.

Take for example [`World::on_destroyed`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/world.rs#L306-L312). While this code doesn't directly dereference the `ctx` pointer, the callback almost certainly will. Since `ctx` is a void pointer, whatever pointer you pass in as `ctx` should be a valid void pointer to an instance of the type the callback is expecting when it dereferences `ctx`. But since you can construct a `*mut c_void` in fully safe code, you can cause undefined behavior just by passing in the wrong type:
```rust
let mut state = 1u8;
let state_ptr = &mut state as *mut _ as *mut c_void; // fully safe c_void pointer construction
let world = World::new();
world.on_destroyed(some_callback_that_doesnt_expect_a_u8, state_ptr);
drop(world); // bad things happen
```

[`EcsSerializer::value_id`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/addons/meta/mod.rs#L199C8-L199C16) Same as `World::on_destroyed`

[`World::to_expr_id`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/addons/script/mod.rs#L90C12-L90C22) Same as `World::on_destroyed` I think?

[`Script::to_expr_id`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/addons/script/unmanaged_script.rs#L216) Same as `World::on_destroyed` I think?

---

[`register_ids_descriptor_at`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/query_tuple.rs#L395C8-L395C34) You can pass in a null pointer as `world` which is then passed to `NotNull::new_unchecked` without any checks, causing UB. Even with a not null check I wouldn't be suprised if there's more bad things that could happen if you passed in random pointers.

---

[`create_tuple_with_row`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/query_tuple.rs#L470) Another thing you can do in fully safe code is manufacture a pointer of any type, ie `let totally_an_ecs_iter_t = &literally_anything as *const _ as *const sys::ecs_iter_t`, which you can then pass to `create_tuple_with_row`, which blindly passes it to `ecs_field_at`, and then everything explodes.

[`ecs_has_pair`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/utility/functions.rs#L84C8-L84C20) Same thing, you can safely create a `*const sys::ecs_world_t` that doesn't actually contain an `ecs_world_t` and break everything.

[`ecs_add_pair`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/utility/functions.rs#L101C8-L101C20) Same as `ecs_has_pair`

[`return_type_for_tuple`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/cloned_tuple.rs#L227) Same as `ecs_has_pair` I think? More unsure on this one

[`return_type_for_tuple`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/get_tuple.rs#L398C8-L398C29) Same as the other `return_type_for_tuple`

---

[`register_lifecycle_actions`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/components/lifecycle_traits.rs#L70C8-L70C34) Not sure why this is marked, the allow should probably be removed (that's another thing, switching from allows to expects, but that's a whole different issue)

[`is_registered_with_world`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/component_registration/registration_traits.rs#L359C8-L359C32) Same as `register_lifecycle_actions`

[`EntityView::new`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/entity_view/entity_view_const.rs#L178) Same as `register_lifecycle_actions`

[`Script::to_expr`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/addons/script/unmanaged_script.rs#L239) Same as `register_lifecycle_actions`

[`EntityView::clear`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/entity_view/entity_view_mut.rs#L1730) Same as `register_lifecycle_actions`

---

[`Cursor::new`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/addons/meta/cursor.rs#L13) I'm unsure about this one. Just reading the description of `ecs_meta_cursor`, I don't understand what the cursor actually does with the void pointer, so I'm not sure if it's ever derefed/could lead to UB if you pass in bad values.

---

[`World::set_context`](https://github.com/Indra-db/Flecs-Rust/blob/f2762ad0594b1efa8603f311d3cee7507ebf4166/flecs_ecs/src/core/world.rs#L1148) This might be the only one that's actually correct to allow `not_unsafe_ptr_arg_deref`. As long as `ecs_set_ctx`/`ecs_get_ctx` don't explode on null pointers/don't deref the `*mut c_void` (I don't see why they would need to), then this should be fine, since you can only ever get the same `*mut c_void` back and it isn't dereferenced anywhere along the way.

---

As a final note, going back to the "other thing" from `register_lifecycle_actions`, I'm not sure what the MSRV is for this project, but if its >= 1.81.0 you should be using the changes from lint reasons: instead of `#[allow(lint)]`, switching to `#[expect(lint, reason="reason here")]` both makes sure that you document why the lint is ignored, and will give a warning if you ever fix the underlying reason for ignoring it in the first place.
